compiler/tests-spec/testData/diagnostics/linked/expressions/not-null-assertion-expression/p-3/pos/1.1.fir.kt

```diff
@@ -23,15 +23,15 @@ import checkSubtype
 fun case1() {
     val a = JavaClass.STR
     <!DEBUG_INFO_EXPRESSION_TYPE("kotlin.String..kotlin.String?! & kotlin.String")!>a<!>
-    val res = a!!
+    val res = a<!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!>
```

Comment:
This and below are wrong, but as shown in the debug info above, smartcast info `String` caused that. Not sure where it came from, yet.

Reply:
I think this is also related to https://github.com/JetBrains/kotlin/pull/4265#discussion_r605396082 (or just one above). Those are all platform types without annotations (like `@NotNull` or `@Nullable`), which are flexible types. But then, incorrectly smartcasted after computing intersection of itself: `it(ft(J..J?), ft(J..J?))` => `J` ?!

I'm not sure the fix should be: 1) `it(X, X)` => `X` or 2) not adding `X` in the beginning if types from smartcast info contains itself: https://github.com/JetBrains/kotlin/blob/master/compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/ResolveUtils.kt#L253 @demiurg906 any recommentations?

Reply:
Yep, `ConeTypeIntersector` must be fixed, so `(A, A?) & (A, A?) = (A, A?)`. Looks like to fix it it's enough to add `this is ConeFlexibleType -> UNKNOWN` branch to `resultNullability` method in `ConeTypeIntersector` (line 145)

Reply:
On top of https://github.com/JetBrains/kotlin/pull/4269, it's now `ft(String..String?)`, and wrong warnings are gone.